### PR TITLE
[Serve] Add proxy routers

### LINKED ISSUE #38193: [Serve] Productionize gRPC proxy work
### Description

We have explored all options and finalized interface for using the new gRPC proxy. Let's clean up the code and make sure all parts are tested. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

- Moved `LongestPrefixRouter` to `proxy_router.py` and created a base class `ProxyRouter`
- Added `EndpointRouter` to match endpoints (will be use in gRPC Proxy)
- Renamed  `test_http_prefix_matching.py` to `test_proxy_router.py` and added tests for `EndpointRouter`
- Added new argument `proxy_router_class` on `GenericProxy` and use `LongestPrefixRouter` for `HTTPProxy`

## Related issue number

Second PR to close https://github.com/ray-project/ray/issues/38193

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -159,6 +159,25 @@ def update_routes(self, endpoints: Dict[EndpointTag, EndpointInfo]) -> None:
         self.route_info = route_info
         self.app_to_is_cross_language = app_to_is_cross_language
 
+    def match_target(self, target: str) -> Optional[str]:
+        """Return the route from the target.
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:163]: What is a target and how does it differ from a route?

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:163]: So `self.route_info` looks something like this
```
{
    '/': ('app1_http-deployment', 'app1'), 
    '/foo': ('app2_http-deployment', 'app2'), 
    '/bar': ('app3_grpc-deployment', 'app3'), 
    '/baz': ('app4_grpc-deployment-model-composition', 'app4')
}
```

`route` is the the keys such as `/`, `/foo`. `target` is matching with the names such as `app1_http-deployment` and `app2_http-deployment`

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:163]: I can change `target` to `target_endpoint` if that clear things a bit, but I feel either name is not the best description of what it does 😅

**Comment by GeneDer** [comment]
@edoakes Can you please take a look at this one when you have a sec so we can raise the following gRPC Proxy PR🙏

File: python/ray/serve/_private/proxy_router.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        # Routes sorted in order of decreasing length.
+        self.sorted_routes: List[str] = list()
+        # Endpoints associated with the routes.
+        self.route_info: Dict[str, EndpointTag] = dict()
+        # Contains a ServeHandle for each endpoint.
+        self.handles: Dict[EndpointTag, RayServeHandle] = dict()
+        # Map of application name to is_cross_language.
+        self.app_to_is_cross_language: Dict[ApplicationName, bool] = dict()
+
+    def endpoint_exists(self, endpoint: EndpointTag) -> bool:
```
Comment (sihanwang41) [inline python/ray/serve/_private/proxy_router.py:48]: is this function being used anywhere?

File: python/ray/serve/_private/proxy_router.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self, target_route: str
+    ) -> Optional[Tuple[str, RayServeHandle, str, bool]]:
+        """Return the endpoint match among existing routes for the route.
+        Args:
+            target_route: endpoint to match against.
+        Returns:
+            (route, handle, app_name, is_cross_language) if found, else None.
+        """
+        for endpoint_tag, handle in self.handles.items():
+            if target_route == str(endpoint_tag):
```
Comment (sihanwang41) [inline python/ray/serve/_private/proxy_router.py:180]: endpoint_tag is already a string?

File: python/ray/serve/_private/proxy_router.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:48]: Oh good point, this whole class is just moved into this new file and added a base class. I did a code search and doesn't seem like it's being used currently. Will just drop it

File: python/ray/serve/_private/proxy_router.py (hunk 3 above)
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:180]: So with Cindy's recent change `endpoint_tag` is an object [`DeploymentID`](https://github.com/ray-project/ray/blob/f6930684b469cba16f34a016a0a06e1653de4019/python/ray/serve/_private/common.py#L33) We need to call `str()` to get the concatenated name for string comparison

Comment (sihanwang41) [reply python/ray/serve/_private/proxy_router.py:180]: Oh, thanks for checking, I am fooled by the annotations, can you update it correspondingly?

Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:177]: Which annotations are you talking about? 😅

Comment (sihanwang41) [reply python/ray/serve/_private/proxy_router.py:177]: nvm, I was checking the old master...

Comment (sihanwang41) [reply python/ray/serve/_private/proxy_router.py:177]: Thanks for checking!

File: python/ray/serve/_private/proxy_router.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                f"Deleting {len(existing_handles)} unused handles.",
+                extra={"log_to_stderr": False},
+            )
+        for endpoint in existing_handles:
+            del self.handles[endpoint]
+
+    def match_route(
+        self, target_route: str
+    ) -> Optional[Tuple[str, RayServeHandle, str, bool]]:
+        """Return the endpoint match among existing routes for the route.
```
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:170]: The concept of route and router came from `LongestPrefixRouter`, do you have any suggestion for the new names? 😅

[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:170]: Maybe just `HandleFinder`?

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/proxy_router.py:170]: "get handle for endpoint"

[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py:170]: hmm but then that diverges on what http is doing. We need to have shared interface between the proxies. `update_routes` and `match_route` are abstract methods

[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/proxy_router.py]: Talked offline, will just use different method for different proxies, and dropping the abstract `match_route` method

**Review by edoakes — APPROVED** [review]

**Pull request merged**